plugins/action/win_pssession_configuration.py

```diff
@@ -0,0 +1,110 @@
+__metaclass__ = type
+
+import ntpath
+import time
+
+from ansible.errors import AnsibleError
+from ansible.plugins.action import ActionBase
+from ansible.utils.vars import merge_hash
+from ansible.utils.display import Display
+display = Display()
+
+
+class ActionModule(ActionBase):
+    _default_async_timeout = 300
+    _default_async_poll = 1
+
+    def __init__(self, *args, **kwargs):
+        super(ActionModule, self).__init__(*args, **kwargs)
+
+    def run(self, tmp=None, task_vars=None):
+        self._supports_check_mode = True
+        self._supports_async = True
+        check_mode = self._play_context.check_mode
+        async_timeout = self._task.async_val
+        async_poll = self._task.poll
+
+        # fake out the super so it doesn't stop us from using check mode with async
+        if check_mode:
+            self._task.async_val = 0
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+
+        # build the wait_for_connection object for later use
+        wait_for_connection_task = self._task.copy()
+        wait_for_connection_task.args = {
+            'timeout': async_timeout or self._default_async_timeout,
+            'sleep': async_poll or self._default_async_poll
+        }
+        wait_connection_action = self._shared_loader_obj.action_loader.get(
+            'wait_for_connection',
+            task=wait_for_connection_task,
+            connection=self._connection,
+            play_context=self._play_context,
+            loader=self._loader,
+            templar=self._templar,
+            shared_loader_obj=self._shared_loader_obj
+        )
+
+        # if it's not in check mode, call the module async so the WinRM restart doesn't kill ansible
+        if not check_mode:
+            self._task.async_val = async_timeout or self._default_async_timeout
+
+        result = status = self._execute_module(
+            task_vars=task_vars,
+            module_args=self._task.args
+        )
+        display.vvvv("Internal Async Result: %r" % status)
+
+        # if we're in check mode (not doing async) or not polling, return the result now
+        if check_mode or async_poll == 0:
+            return result
+
+        # remove the async flag so that our calls to async_status and wait_for_connection don't try to run async
+        self._task.async_val = 0
+
+        async_status_args = {
+            'jid': status['ansible_job_id'],
+            '_async_dir': ntpath.dirname(status['results_file'])
+        }
+
+        # Retries here is a fallback in case the module fails in an unexpected way
+        # which can sometimes not properly set the failed field in the return.
+        # It is not related to async retries.
+        # Without this, that situation would cause an infinite loop.
+        max_retries = 3
+        retries = 0
+        while not check_mode:
+            try:
+                # check up on the async job
+                job_status = self._execute_module(
```

Comment:
Why are you calling the async_status module and not the action plugin like you've done for `wait_for_connection`?

Reply:
Hm, I wrote the action plugin in a much more basic version of this module from last year, and I don't remember why that's the case. It might be that I wrote that part first, as it's much more obvious how to call modules than other action plugins. I'll try it out the other way.

Reply:
I figured it out why I couldn't get this to work in the past; I was trying to duplicate the current task object when I should have been creating a new one. I found a good example in `task_executor.py` which set me straight; thanks!